src/bitset.rs

```diff
@@ -0,0 +1,447 @@
+        BitSet {
+            layer3: 0,
+            layer2: Vec::new(),
+            layer1: Vec::new(),
+            layer0: Vec::new(),
+        }
+    }
+
+    #[inline]
+    fn valid_range(max: Index) {
+        if (MAX_EID as u32) < max {
+            panic!("Expected index to be less then {}, found {}", MAX_EID, max);
+        }
+    }
+
+    /// Create an empty BitSet with up to max Index
+    pub fn with_capacity(max: Index) -> BitSet {
+        Self::valid_range(max);
+        let mut value = BitSet::new();
+        value.extend(max);
+        value
+    }
+
+    #[inline(never)]
+    fn extend(&mut self, id: Index) {
+        Self::valid_range(id);
+        let (p0, p1, p2) = offsets(id);
+
+        if self.layer2.len() <= p2 {
+            let count = p2 - self.layer2.len() + 1;
+            self.layer2.extend(repeat(0).take(count));
+        }
+        if self.layer1.len() <= p1 {
+            let count = p1 - self.layer1.len() + 1;
+            self.layer1.extend(repeat(0).take(count));
+        }
+        if self.layer0.len() <= p0 {
+            let count = p0 - self.layer0.len() + 1;
+            self.layer0.extend(repeat(0).take(count));
+        }
+    }
+
+    /// this is used to set the levels in the hierarchy
+    /// when the lowest layer was set from 0
+    fn add_slow(&mut self, id: Index) {
+        let (_, p1, p2) = offsets(id);
+        self.layer1[p1] |= id.mask::<Shift1>();
+        self.layer2[p2] |= id.mask::<Shift2>();
+        self.layer3 |= id.mask::<Shift3>();
+    }
+
+    /// Add `id` to the bitset. Returning if the value was
+    /// already in the set before it was added
+    #[inline]
+    pub fn add(&mut self, id: Index) -> bool {
+        let (p0, mask) = (id.offset::<Shift1>(), id.mask::<Shift0>());
+
+        if p0 >= self.layer0.len() {
+            self.extend(id);
+        }
+
+        if self.layer0[p0] & mask != 0 {
+            return true;
+        }
+
+        // we need to set the bit on every layer to indicate
+        // that the value can be found here.
+        let old = self.layer0[p0];
+        self.layer0[p0] |= mask;
+        if old == 0 {
+            self.add_slow(id);
+        } else {
+            self.layer0[p0] |= mask;
+        }
+        false
+    }
+
+    /// Remove `id` from the set, returns true if the value
+    /// was removed, returns false if the value was not set
+    /// to begin with
+    #[inline]
+    pub fn remove(&mut self, id: Index) -> bool {
+        let (p0, p1, p2) = offsets(id);
+
+        if p0 >= self.layer0.len() {
+            return false;
+        }
+
+        if self.layer0[p0] & id.mask::<Shift0>() == 0 {
+            return false;
+        }
+
+        // if the bitmask was set we need to clear
+        // its bit from layer0 to 3. the layers abover only
+        // should be cleared if the bit cleared was the last bit
+        // in its set
+        self.layer0[p0] &= !id.mask::<Shift0>();
+        if self.layer0[p0] != 0 {
```

Comment:
Since you've already removed this bit, how do you know if it was there in the first place (in order to `return true`)?